### PR TITLE
[DMP-748] Remove items from offers on Offers and Payment

### PR DESCRIPTION
**Pull request by luisdeacero**
## What is done

[[DMP-748] [Frontend] Remover Productos de Oferta Seleccionada](https://deacero.atlassian.net/browse/DMP-748)

## How is done

1. Added a handleRemove function recieving a ArticileDetail param
2. Call handleRemove depending on clicked delete item
3. Use setters for context to update offer
4. Couldn't figure out why the component wasn't rerendering when an item was deleted, so added a random seed to force an update (Refactor Pending)

## How to test

1. Run the project via `npm start`
2. Add several products to cart (3 is more than enough)
3. Head over to cart
4. Select an offer
5. Try to remove an item
6. Item should disappear and price should update
7. Head over to payment
8. Try to remove an item
9. Item should disappear and price should update

## Evidence

Add images or screen  recordings on how the feature was working after and before the fix (if applies)

### CODE REVIEW

**Comment by markriverac** [comment]
I just noticed the "help wanted" label. Please reach us to talk about what your problem is :)

File: src/views/Offers/OfferDetail/OfferDetail.tsx [hunk 1]
```diff
@@ -49,6 +53,32 @@ const OfferDetail: React.FC = () => {
...
+      auxSelectedOffer.total -=
+        offerArticles[index].price * offerArticles[index].quantity;
+      aux = offerArticles;
+      aux.splice(index, 1);
+      setSelectedOffer(auxSelectedOffer);
+      setOfferArticles(aux);
+    }
+    console.log(seed);
+    reset();
+  }
```
Comment (markriverac) [inline src/views/Offers/OfferDetail/OfferDetail.tsx:80]: I propose a refactoring like this:
```javascript
 function handleRemove(item: ArticleDetail) {
    const itemIndexInOfferArticles = offerArticles.indexOf(item);

    if (itemIndexInOfferArticles < 0) return;

    const selectedOfferArticles = selectedOffer.shipments[0].articles;
    let itemIndexInSelectedOfferArticles = -1;
    selectedOfferArticles.find((art) => {
      itemIndexInSelectedOfferArticles++;
      return art.articleSKU === Number(item.article_id);
    });

    const selectedOfferCopy = { ...selectedOffer };
    const offerArticlesCopy = [...offerArticles];
    const newArticles = [...selectedOfferArticles];

    newArticles.splice(itemIndexInSelectedOfferArticles, 1);
    selectedOfferCopy.shipments[0].articles = newArticles;
    selectedOfferCopy.total -= item.price * item.quantity;
    setSelectedOffer(selectedOfferCopy);

    offerArticlesCopy.splice(itemIndexInOfferArticles, 1);
    setOfferArticles(offerArticlesCopy);
  }
```
I think it works, but I didn't test it for all scenarios. If it doesn't work, at least try to use some of it, like the renaming of variables and cleaner code blocks

**Review by markriverac — CHANGES_REQUESTED** [review]
Please try the refactoring I proposed.
Also, there are a few things I noticed.
1. You can't remove substitute products, so I recommend verifying first if the articles have the `substitutes ` param which indicates what product is being substituted.
2. I don't know if this is just because of my refactoring, but moving from one offer to another doesn't correctly resets the products sometimes.
3. If I remove all the items from the selected offer, what should happen? I can click on "next" with an empty offer. Please reach the UX team for this.

File: src/views/Offers/OfferDetail/OfferDetail.tsx (hunk 1 above)
Comment (luisdeacero) [reply src/views/Offers/OfferDetail/OfferDetail.tsx:80]: The refactoring did work, didn't even need to make changes for the cases I tested.

1. Done. You can now remove products aimed to substitute another one.
2. I am not quite sure why that might be, I can make a new bug ticket to attend it.
3. Let me check with them, should I apply the solution when given here or on a new ticket?

File: src/views/Offers/OfferDetail/OfferDetail.tsx [hunk 2]
```diff
@@ -55,28 +51,30 @@ const OfferDetail: React.FC = () => {
...
-        offerArticles[index].price * offerArticles[index].quantity;
-      aux = offerArticles;
-      aux.splice(index, 1);
-      setSelectedOffer(auxSelectedOffer);
-      setOfferArticles(aux);
-    }
-    console.log(seed);
-    reset();
+    const itemIndexInOfferArticles = offerArticles.indexOf(item);
+    console.log(itemIndexInOfferArticles);
```
Comment (markriverac) [inline src/views/Offers/OfferDetail/OfferDetail.tsx:55]: Just remove this console.log, please

File: src/views/Payment/Payment.tsx [hunk 3]
```diff
@@ -130,28 +125,31 @@ const Payment: React.FC = () => {
...
-      aux = offerArticles;
-      aux.splice(index, 1);
-      setSelectedOffer(auxSelectedOffer);
-      setOfferArticles(aux);
-    }
-    console.log(seed);
-    reset();
+    const itemIndexInOfferArticles = offerArticles.indexOf(item);
+
+    console.log(itemIndexInOfferArticles);
```
Comment (markriverac) [inline src/views/Payment/Payment.tsx:130]: This console.log too

**Review by markriverac — APPROVED** [review]
Just remove the console.log :)
Thank you for the update!

**Pull request merged**